File: WebMock/src/core/hint.ts
// ヒント対象のピース選定。Three.js に依存しない純粋なロジック。
// ヒントは「選ばれたピースを解答の位置・向きへ置いて固定する」機能で、ここはその対象選びだけを担う。

import { equalsVec3 } from './grid';
import type { Placement } from './piece';

/** 配置が解答と完全に一致しているか（位置と向きの両方）。 */
function matchesSolution(placement: Placement, solution: Placement): boolean {
  return (
    placement.orientation === solution.orientation &&
    equalsVec3(placement.position, solution.position)
  );
}

/**
 * ヒントで正解位置へ送るピースの id を選ぶ。使えるヒントが無ければ null。
 *
 * - 未固定（lockedIds に含まれない）のピースが 1 個以下なら null。
 *   最後の 1 ピースをヒントで埋めると操作せずクリアできてしまうため（要望の制約）。
 * - 未固定のうち、現在の配置が解答と異なるものの中で id が最小のものを選ぶ。
 * - 未固定がすべて解答と一致していれば、未固定の id が最小のものを選ぶ。
 *
 * 解釈: 選び方は要望に指定が無いので、上のように決定的な規則にする（同じ盤面では常に同じヒント）。
 */
export function pickHintPiece(
  placements: readonly Placement[],
  solution: readonly Placement[],
  lockedIds: Iterable<number>,
): number | null {
  const solutionById = new Map<number, Placement>();
  for (const placement of solution) solutionById.set(placement.pieceId, placement);

  const locked = new Set<number>(lockedIds);
  let fallback: number | null = null;
  let target: number | null = null;
  let unlockedCount = 0;

  for (const placement of placements) {
    const answer = solutionById.get(placement.pieceId);
    if (answer === undefined) {
      throw new Error(`ヒント: 解答に無いピース id ${placement.pieceId}`);
    }
    if (locked.has(placement.pieceId)) continue;
    unlockedCount++;
    if (fallback === null || placement.pieceId < fallback) fallback = placement.pieceId;
    if (matchesSolution(placement, answer)) continue;
    if (target === null || placement.pieceId < target) target = placement.pieceId;
  }

  // 最後の 1 ピースはヒントを使えない（未固定 0 個 / 1 個は null）
  if (unlockedCount <= 1) return null;
  return target ?? fallback;
}

File: Source/CUBELITHCore/Private/Hint.cpp
#include "Hint.h"

#include "Grid.h"

namespace Cubelith
{
	namespace
	{
		/** 配置が解答と完全に一致しているか（位置と向きの両方）。TS の matchesSolution */
		bool MatchesSolution(const FPlacement& Placement, const FPlacement& Solution)
		{
			return Placement.Orientation == Solution.Orientation && EqualsVec3(Placement.Position, Solution.Position);
		}
	}

	TOptional<int32> PickHintPiece(
		TArrayView<const FPlacement> Placements, TArrayView<const FPlacement> Solution, TArrayView<const int32> LockedIds)
	{
		// 解答を id 引きの表にする（TS の solutionById）。引くだけで走査しないので、
		// TMap の反復順が結果に効くことはない（Docs/SPEC_UE.md 7.1）
		TMap<int32, const FPlacement*> SolutionById;
		SolutionById.Reserve(Solution.Num());
		for (const FPlacement& Answer : Solution)
		{
			SolutionById.Add(Answer.PieceId, &Answer);
		}

		// 「含むかどうか」だけを見る集合なので TSet でよい（同 7.1）
		TSet<int32> Locked;
		Locked.Reserve(LockedIds.Num());
		for (const int32 LockedId : LockedIds)
		{
			Locked.Add(LockedId);
		}

		TOptional<int32> Fallback;
		TOptional<int32> Target;
		int32 UnlockedCount = 0;

		for (const FPlacement& Placement : Placements)
		{
			const FPlacement* const* Answer = SolutionById.Find(Placement.PieceId);
			// 解答に無い id は呼び出し側のバグ（TS は Error）
			checkf(Answer != nullptr, TEXT("ヒント: 解答に無いピース id %d"), Placement.PieceId);

			if (Locked.Contains(Placement.PieceId))
			{
				continue;
			}

			++UnlockedCount;
			if (!Fallback.IsSet() || Placement.PieceId < Fallback.GetValue())
			{
				Fallback = Placement.PieceId;
			}
			if (MatchesSolution(Placement, **Answer))
			{
				continue;
			}
			if (!Target.IsSet() || Placement.PieceId < Target.GetValue())
			{
				Target = Placement.PieceId;
			}
		}

		// 最後の 1 ピースはヒントを使えない（未固定 0 個 / 1 個は未設定。RULES.md 3.7）
		if (UnlockedCount <= 1)
		{
			return TOptional<int32>();
		}
		return Target.IsSet() ? Target : Fallback;
	}
}
